engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,122 @@
+# PLAN-0002: Ethereum Tracing Cache
+
+<dl>
+  <dt>Author</dt>
+  <dd>Zachary Burns</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0002-ethereum-tracing-cache.md">RFC-0002 Ethereum Tracing Cache</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/9">https://github.com/graphprotocol/rfcs/pull/9</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>None</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-20</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Summary
+
+Implements RFC-0002: Ethereum Tracing Cache
+
+## Implementation
+
+These changes happen within `ethereum_adapter.rs`, `store.rs` and `db_schema.rs`.
+
+### ServiceCache trait
+EthereumCallCache will be renamed to ServiceCache and have roughly this signature:
+```rust
+trait ServiceCache {
```

Comment:
I get the sentiment of wanting a more reusable cache, but I feel there is a benefit to the self-documenting schema of proper Postgres tables and a more specific API. What I do like here is the switch to futures and the callback to perform work on a cache miss.

We may not use this information yet, but the `block_number` and `contract_address` may become important debugging and cache cleanup tools. Opaque byte data lacks hints like that and, like with Ethereum's byte encoding, requires the caller to know what the structure of the data is, rather than having it manifested in a schema.

I'd prefer something like

```rust
struct EthCallKey {
    contract_address: ethabi::Address,
    block: EthereumBlockPointer,
    encoded_call: &[u8],
}

struct EthCallResult {
    contract_address: ethabi::Address,
    block: EthereumBlockPointer,
    encoded_call: &[u8],
    return_value: &[u8],
}
  
trait EthereumCache {
    fn get_eth_call(
        &self,
        key: EthCallKey,
        f: impl Future<Output = Result<EthCallResult>, Error>>,
    ) -> Box<dyn Future<Output = Result<EthCallResult, Error>>>;

    fn get_trace_filter(
        &self,
        key: TraceFilterKey,
        f: impl Future<Output = Result<TraceFilterResult>, Error>>,
    ) -> Box<dyn Future<Output = Result<TraceFilterResult, Error>>>;
}
```

Plus `try_store_eth_call` / `try_get_eth_call`, `try_store_trace_filter` / `try_get_trace_filter` methods.

That way we can keep the `eth_call_data` table and just add an `eth_trace_filter_data` table alongside it.

Reply:
(Note sure the type names are the best but you get the idea.)

Reply:
At the same time, I see that the last accessed hint may be sufficient for cleaning up the cache periodically. As you can tell, I'm torn. 😄